File: app/src/main/java/com/arseniy/maps/activities/AddCityActivity.java
package com.arseniy.maps.activities;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Geocoder;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

import com.arseniy.maps.R;
import com.arseniy.maps.data.MapData;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.maps.model.LatLng;

import java.io.IOException;
import java.util.List;

public class AddCityActivity extends AppCompatActivity {

    private EditText cityNameField;
    private EditText cityLatField;
    private EditText cityLngField;
    private Geocoder coder;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.add_city);

        onMapBtnInit();

        cityNameField = findViewById(R.id.cityNameField);
        cityLatField = findViewById(R.id.cityLatField);
        cityLngField = findViewById(R.id.cityLngField);

        addCityBtnInit();
        coder = new Geocoder(this);
        getLatLngFromApiBtnInit();

        getFromCurLocBtnInit();
    }

    @Override
    protected void onResume() {
        super.onResume();
        cityNameField.requestFocus();
    }

    private void onMapBtnInit() {
        Button onMapBtn = findViewById(R.id.onMapBtn);
        onMapBtn.setOnClickListener(view -> callOnMapPick());
    }

    private void  callOnMapPick() {
        Intent intent = new Intent(this, PickLocationOnMapActivity.class);
        try {
            final double cityLat = Double.parseDouble(cityLatField.getText().toString());
            final double cityLng = Double.parseDouble(cityLngField.getText().toString());
            intent.putExtra("startingLat", cityLat);
            intent.putExtra("startingLng", cityLng);
        } catch (Exception ignored) {
        }

        startActivityForResult(intent, 0);
    }

    @SuppressLint("SetTextI18n") //that will add commas instead of dots.
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == resultCode) {
            final float lat = data.getFloatExtra("lat", 0f);
            final float lng = data.getFloatExtra("lng", 0f);
            setLatLntData(new LatLng(lat, lng));
        }
    }

    private void addCityBtnInit() {
        Button addCityBtn = findViewById(R.id.addCityBtn);
        addCityBtn.setOnClickListener(view -> {
            try {
                final String cityName = cityNameField.getText().toString().trim();
                final float cityLat = Float.parseFloat(cityLatField.getText().toString());
                final float cityLng = Float.parseFloat(cityLngField.getText().toString());
                storeCity(cityName, cityLat, cityLng);
                this.finish();
            } catch (Exception e) {
                setResult(-1);
                Toast.makeText(this, "Ошибка создания записи, проверьте данные.", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void storeCity(String name, float lat, float lng) {
        MapData.addCityToStorage(name, lat, lng); //sharedPrefs
        setResult(0, new Intent().putExtra("cityName", name)); //list
    }

    @SuppressLint("SetTextI18n") //that will add commas instead of dots.
    private void getLatLngFromApiBtnInit() {
        Button getLatLngFromApiBtn = findViewById(R.id.getLatLngFromApiBtn);
        getLatLngFromApiBtn.setOnClickListener(view -> {
            LatLng pos = getLocationFromAddress(cityNameField.getText().toString().trim());
            if (pos != null) {
                setLatLntData(pos);
                callOnMapPick();
            }
        });
    }

    private LatLng getLocationFromAddress(String strAddress) {
        try {
            List<Address> address = coder.getFromLocationName(strAddress, 1);
            if (address != null && address.size() > 0) {
                Address location = address.get(0);
                return new LatLng(location.getLatitude(), location.getLongitude());
            }
        } catch (IOException ignored) {
        }

        Toast.makeText(this, "Нет интернет-соединения или такой город не найден.", Toast.LENGTH_SHORT).show();
        return null;
    }

    private void getFromCurLocBtnInit() {
        Button fromCurLocBtn = findViewById(R.id.fromCurLocBtn);
        fromCurLocBtn.setOnClickListener(view -> {
            FusedLocationProviderClient mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this);
            if (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 0);
            }
            mFusedLocationClient.getLastLocation().addOnSuccessListener(result -> setLatLntData(new LatLng(result.getLatitude(), result.getLongitude())));
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions, @NonNull int[] grantResults) {
        // If request is cancelled, the result arrays are empty.
        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            FusedLocationProviderClient mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this);
            if (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                mFusedLocationClient.getLastLocation().addOnSuccessListener(result -> setLatLntData(new LatLng(result.getLatitude(), result.getLongitude())));
            }
        } else {
            Toast.makeText(this, "Неудалось получить локацию - разрешение не выдано.", Toast.LENGTH_SHORT).show();
        }
    }

    private void setLatLntData(LatLng pos) {
        if (pos != null) {
            String lat = Double.toString(pos.latitude);
            if (lat.length() > 7)
                lat = lat.substring(0, 7);
            cityLatField.setText(lat);
            String lng = Double.toString(pos.longitude);
            if (lng.length() > 7)
                lng = lng.substring(0, 7);
            cityLngField.setText(lng);
        }
    }
}

File: app/src/main/java/com/arseniy/maps/activities/AddCityActivity.kt
package com.arseniy.maps.activities

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Geocoder
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat

import com.arseniy.maps.R
import com.arseniy.maps.data.MapData
import com.google.android.gms.location.LocationServices
import com.google.android.gms.maps.model.LatLng

import java.io.IOException

class AddCityActivity : AppCompatActivity() {

    private var cityNameField: EditText? = null
    private var cityLatField: EditText? = null
    private var cityLngField: EditText? = null
    private var coder: Geocoder? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.add_city)

        cityNameField = findViewById(R.id.cityNameField)
        cityLatField = findViewById(R.id.cityLatField)
        cityLngField = findViewById(R.id.cityLngField)

        coder = Geocoder(this)

        onMapBtnInit()
        addCityBtnInit()
        getLatLngFromApiBtnInit()
        getFromCurLocBtnInit()
    }

    override fun onResume() {
        super.onResume()
        cityNameField!!.requestFocus()
    }

    private fun onMapBtnInit() {
        val onMapBtn = findViewById<Button>(R.id.onMapBtn)
        onMapBtn.setOnClickListener { callOnMapPick() }
    }

    private fun callOnMapPick() {
        val intent = Intent(this, PickLocationOnMapActivity::class.java)
        try {
            val cityLat = java.lang.Double.parseDouble(cityLatField!!.text.toString())
            val cityLng = java.lang.Double.parseDouble(cityLngField!!.text.toString())
            intent.putExtra("startingLat", cityLat)
            intent.putExtra("startingLng", cityLng)
        } catch (ignored: Exception) { }

        startActivityForResult(intent, 0)
    }

    @SuppressLint("SetTextI18n") //that will add commas instead of dots.
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == resultCode) {
            val lat = data!!.getFloatExtra("lat", 0f)
            val lng = data.getFloatExtra("lng", 0f)
            setLatLntData(LatLng(lat.toDouble(), lng.toDouble()))
        }
    }

    private fun addCityBtnInit() {
        val addCityBtn = findViewById<Button>(R.id.addCityBtn)
        addCityBtn.setOnClickListener {
            try {
                val cityName = cityNameField!!.text.toString().trim { it <= ' ' }
                val cityLat = java.lang.Float.parseFloat(cityLatField!!.text.toString())
                val cityLng = java.lang.Float.parseFloat(cityLngField!!.text.toString())
                storeCity(cityName, cityLat, cityLng)
                this.finish()
            } catch (e: Exception) {
                setResult(-1)
                Toast.makeText(this, "Ошибка создания записи, проверьте данные.", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun storeCity(name: String, lat: Float, lng: Float) {
        MapData.addCityToStorage(name, lat.toDouble(), lng.toDouble()) //sharedPrefs
        setResult(0, Intent().putExtra("cityName", name)) //list
    }

    @SuppressLint("SetTextI18n") //that will add commas instead of dots.
    private fun getLatLngFromApiBtnInit() {
        val getLatLngFromApiBtn = findViewById<Button>(R.id.getLatLngFromApiBtn)
        getLatLngFromApiBtn.setOnClickListener {
            val pos = getLocationFromAddress(cityNameField!!.text.toString().trim { it <= ' ' })
            if (pos != null) {
                setLatLntData(pos)
                callOnMapPick()
            }
        }
    }

    private fun getLocationFromAddress(strAddress: String): LatLng? {
        try {
            val address = coder!!.getFromLocationName(strAddress, 1)
            if (address != null && address.size > 0) {
                val location = address[0]
                return LatLng(location.latitude, location.longitude)
            }
        } catch (ignored: IOException) { }

        Toast.makeText(this, "Нет интернет-соединения или такой город не найден.", Toast.LENGTH_SHORT).show()
        return null
    }

    private fun getFromCurLocBtnInit() {
        val fromCurLocBtn = findViewById<Button>(R.id.fromCurLocBtn)
        fromCurLocBtn.setOnClickListener {
            val mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
            if (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 0)
            }
            mFusedLocationClient.lastLocation.addOnSuccessListener { result -> setLatLntData(LatLng(result.latitude, result.longitude)) }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>, grantResults: IntArray) {
        // If request is cancelled, the result arrays are empty.
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            val mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
            if (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                mFusedLocationClient.lastLocation.addOnSuccessListener { result -> setLatLntData(LatLng(result.latitude, result.longitude)) }
            }
        } else {
            Toast.makeText(this, "Неудалось получить локацию - разрешение не выдано.", Toast.LENGTH_SHORT).show()
        }
    }

    private fun setLatLntData(pos: LatLng?) {
        if (pos != null) {
            cityLatField!!.setText(String.format("%.4f", pos.latitude))
            cityLngField!!.setText(String.format("%.4f", pos.longitude))
        }
    }
}
